libs/cmdio/render.go

```diff
@@ -56,12 +178,94 @@ func renderJson(w io.Writer, v any) error {
 		return err
 	}
 	_, err = w.Write([]byte("\n"))
-	return err
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d defaultRenderer) renderTemplate(_ context.Context, t *template.Template, w *tabwriter.Writer) error {
+	return t.Execute(w, d.it)
+}
+
+// Returns something implementing one of the following interfaces:
+//   - jsonRenderer
+//   - textRenderer
+//   - templateRenderer
+func newRenderer(it any) any {
+	if r, ok := it.(io.Reader); ok {
+		return readerRenderer{reader: r}
+	}
+	if iterator, ok := newReflectIterator(it); ok {
+		return iteratorRenderer{t: iterator}
+	}
+	return defaultRenderer{it: it}
 }
 
-func renderTemplate(w io.Writer, tmpl string, v any) error {
+type bufferedFlusher struct {
+	w io.Writer
+	b bytes.Buffer
+}
+
+func (b bufferedFlusher) Write(bs []byte) (int, error) {
+	return b.b.Write(bs)
+}
+
+func (b bufferedFlusher) Flush() error {
+	_, err := b.Write(b.b.Bytes())
+	if err != nil {
+		return err
+	}
+	b.b.Reset()
+	return nil
+}
+
+func newBufferedFlusher(w io.Writer) writeFlusher {
+	return bufferedFlusher{w: w}
+}
+
+func renderWithTemplate(r any, ctx context.Context, outputFormat flags.Output, w io.Writer, headerTemplate, template string) error {
+	// TODO: add terminal width & white/dark theme detection
+	switch outputFormat {
+	case flags.OutputJSON:
+		if jr, ok := r.(jsonRenderer); ok {
+			return jr.renderJson(ctx, newBufferedFlusher(w))
+		}
+		return errors.New("json output not supported")
+	case flags.OutputText:
+		if tr, ok := r.(templateRenderer); ok && template != "" {
+			return renderUsingTemplate(ctx, tr, w, headerTemplate, template)
+		}
+		if tr, ok := r.(textRenderer); ok {
+			return tr.renderText(ctx, newBufferedFlusher(w))
```

Comment:
This may cause a regression; if we're rendering a reader, use of the buffered flusher will cache the full payload in memory before writing anything. For large files on UC volumes this is potentially gigabytes.

Also see the comments in the godoc of `io.Copy` ([here](https://pkg.go.dev/io#Copy)): if the src implements `WriterTo` or the dst implements `ReaderFrom`, the copy is short-circuited. By wrapping the underlying writer (which presumably is equal to the actual stdout), the dst no longer implements this interface.